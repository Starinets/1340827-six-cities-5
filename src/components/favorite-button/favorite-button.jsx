import React from 'react';
import * as Type from '../../types';
import {connect} from 'react-redux';

import {redirectToRoute} from '../../store/action';
import {getAuthorizationStatus} from '../../store/selectors';
import {setOfferStatus} from '../../store/api-actions';
import {
  AuthorizationStatus,
  APIRoute,
  FavoriteStatus
} from '../../constants';

class FavoriteButton extends React.PureComponent {

  constructor(props) {
    super(props);

    this.handleBookmarkClick = this.handleBookmarkClick.bind(this);
  }

  handleBookmarkClick() {
    const {
      isFavorite,
      authorizationStatus,
      id,
      redirectToLoginPage,
      changeOfferStatus
    } = this.props;

    if (authorizationStatus !== AuthorizationStatus.AUTH) {
      return redirectToLoginPage;
    }

    return () => {
      return changeOfferStatus(id, isFavorite ? FavoriteStatus.NO : FavoriteStatus.YES);
    };
  }

  render() {
    const {
      className,
      buttonWidth,
      buttonHeight,
      isFavorite
    } = this.props;

    return (
      <button
        onClick={ this.handleBookmarkClick() }
        className={`${ className }__bookmark-button ${ isFavorite ? `${ className }__bookmark-button--active` : `` } button`}
        type="button"
      >
        <svg
          className={ `${ className }__bookmark-icon` }
          width={ buttonWidth }
          height={ buttonHeight }>
          <use xlinkHref="#icon-bookmark"></use>
        </svg>
        <span className="visually-hidden">To bookmarks</span>
      </button>
    );
  }
}

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state)
});

const mapDispatchToProps = (dispatch) => ({
  redirectToLoginPage() {
    dispatch(redirectToRoute(APIRoute.LOGIN));
  },
  changeOfferStatus(id, status) {
    dispatch(setOfferStatus(id, status));
  }
});

FavoriteButton.propTypes = {
  isFavorite: Type.BOOLEAN.isRequired,
  className: Type.STRING.isRequired,
  buttonWidth: Type.NUMBER.isRequired,
  buttonHeight: Type.NUMBER.isRequired,
  authorizationStatus: Type.STRING.isRequired,
  id: Type.NUMBER.isRequired,
  redirectToLoginPage: Type.FUNCTION.isRequired,
  changeOfferStatus: Type.FUNCTION.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(FavoriteButton);
