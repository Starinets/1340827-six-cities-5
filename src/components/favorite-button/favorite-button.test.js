import React from 'react';
import renderer from 'react-test-renderer';
import {FavoriteButton} from './favorite-button';
import {
  emptyFunction,
  FavoriteButtonProperty,
  AuthorizationStatus
} from '../../test-data/test-data';

describe(`Component rendered correctly -> FavoriteButton`, () => {
  it(`the offer is in the favorite list, authorization is AUTH`, () => {
    const tree = renderer.create(
        <FavoriteButton
          isFavorite={ true }
          className={ FavoriteButtonProperty.PlaceCard.CLASS_NAME }
          buttonWidth={ FavoriteButtonProperty.PlaceCard.WIDTH }
          buttonHeight={ FavoriteButtonProperty.PlaceCard.WIDTH }
          authorizationStatus={ AuthorizationStatus.AUTH }
          id={ 1 }
          redirectToLoginPage={ emptyFunction }
          changeOfferStatus={ emptyFunction }
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`the offer isn't in the favorite list, authorization is AUTH`, () => {
    const tree = renderer.create(
        <FavoriteButton
          isFavorite={ false }
          className={ FavoriteButtonProperty.PlaceCard.CLASS_NAME }
          buttonWidth={ FavoriteButtonProperty.PlaceCard.WIDTH }
          buttonHeight={ FavoriteButtonProperty.PlaceCard.WIDTH }
          authorizationStatus={ AuthorizationStatus.AUTH }
          id={ 1 }
          redirectToLoginPage={ emptyFunction }
          changeOfferStatus={ emptyFunction }
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`the offer is in the favorite list, authorization is NO_AUTH`, () => {
    const tree = renderer.create(
        <FavoriteButton
          isFavorite={ true }
          className={ FavoriteButtonProperty.PlaceCard.CLASS_NAME }
          buttonWidth={ FavoriteButtonProperty.PlaceCard.WIDTH }
          buttonHeight={ FavoriteButtonProperty.PlaceCard.WIDTH }
          authorizationStatus={ AuthorizationStatus.NO_AUTH }
          id={ 1 }
          redirectToLoginPage={ emptyFunction }
          changeOfferStatus={ emptyFunction }
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`the offer isn't in the favorite list, authorization is NO_AUTH`, () => {
    const tree = renderer.create(
        <FavoriteButton
          isFavorite={ false }
          className={ FavoriteButtonProperty.PlaceCard.CLASS_NAME }
          buttonWidth={ FavoriteButtonProperty.PlaceCard.WIDTH }
          buttonHeight={ FavoriteButtonProperty.PlaceCard.WIDTH }
          authorizationStatus={ AuthorizationStatus.NO_AUTH }
          id={ 1 }
          redirectToLoginPage={ emptyFunction }
          changeOfferStatus={ emptyFunction }
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
