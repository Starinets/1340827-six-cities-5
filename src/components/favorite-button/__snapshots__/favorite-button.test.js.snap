// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component rendered correctly -> FavoriteButton the offer is in the favorite list, authorization is AUTH 1`] = `
<button
  className="place-card__bookmark-button place-card__bookmark-button--active button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={18}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;

exports[`Component rendered correctly -> FavoriteButton the offer is in the favorite list, authorization is NO_AUTH 1`] = `
<button
  className="place-card__bookmark-button place-card__bookmark-button--active button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={18}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;

exports[`Component rendered correctly -> FavoriteButton the offer isn't in the favorite list, authorization is AUTH 1`] = `
<button
  className="place-card__bookmark-button  button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={18}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;

exports[`Component rendered correctly -> FavoriteButton the offer isn't in the favorite list, authorization is NO_AUTH 1`] = `
<button
  className="place-card__bookmark-button  button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={18}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;
